examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,1762 @@
+
+                # 5. Get the student model predicted original sample `student_x_0`.
+                student_noise_pred = unet(
+                    noisy_student_input,
+                    student_timesteps,
+                    encoder_hidden_states=prompt_embeds.float(),
+                ).sample
+                student_x_0 = denoiser(student_noise_pred, student_timesteps, noisy_student_input)
+
+                # 6. Sample noise and add it to the student's predicted original sample according to the noise
+                # magnitude at each teacher timestep (that is, run the forward process on the teacher model, but
+                # using `student_x_0` instead of latents sampled from the prior).
+                teacher_noise = torch.randn_like(student_x_0)
+                noisy_teacher_input = noise_scheduler.add_noise(student_x_0, teacher_noise, teacher_timesteps)
+
+                # 7. Sample random guidance scales w ~ U[w_min, w_max] for CFG.
+                w = (args.w_max - args.w_min) * torch.rand((bsz,)) + args.w_min
+                w = w.reshape(bsz, 1, 1, 1)
+                # Move to U-Net device and dtype
+                w = w.to(device=latents.device, dtype=latents.dtype)
+
+                # 8. Get teacher model predicted original sample `teacher_x_0`.
+                with torch.no_grad(), torch.autocast("cuda", dtype=weight_dtype):
+                    teacher_cond_noise_pred = teacher_unet(
+                        noisy_teacher_input.detach(),
+                        teacher_timesteps,
+                        encoder_hidden_states=prompt_embeds,
+                    ).sample
+
+                    teacher_uncond_noise_pred = teacher_unet(
+                        noisy_teacher_input.detach(),
+                        teacher_timesteps,
+                        encoder_hidden_states=uncond_prompt_embeds,
+                    ).sample
+
+                    # Get the teacher's CFG estimate of x_0.
+                    teacher_cfg_noise_pred = w * teacher_cond_noise_pred + (1 - w) * teacher_uncond_noise_pred
+                    teacher_x_0 = denoiser(teacher_cfg_noise_pred, teacher_timesteps, noisy_teacher_input)
+
+                ############################
+                # 9. Discriminator Loss
+                ############################
+                discriminator_optimizer.zero_grad(set_to_none=True)
+
+                # 1. Decode real and fake (generated) latents back to pixel space.
+                # NOTE: the paper doesn't mention this explicitly AFAIK but I think this makes sense since the
+                # pretrained feature network for the discriminator operates in pixel space rather than latent space.
+                unscaled_latents = (1 / vae.config.scaling_factor) * latents
+                unscaled_student_x_0 = (1 / vae.config.scaling_factor) * student_x_0
+                real_image = vae.decode(unscaled_latents).sample
```

Comment:
actually this is not necessary, because we already have access to `pixel_values` from real image.

Reply:
Switched to using `pixel_values` in place of `real_image`.